[[ Bug 19911 ]] Ensure android interface callbacks are on engine thread

livecodeali opened the pull request:
Also document thread affinity syntax in binding string.

livecodeali committed:
[[ Bug 19911 ]] Ensure android interface callbacks are on engine thread
This patch changes the thread affinity specification in the binding
string to use an optional `?thread` clause at the end of the string.

runrevmark commented on engine/src/mblandroidlcb.cpp:
```diff
@@ -55,13 +55,13 @@ extern "C" JNIEXPORT void JNICALL Java_com_runrev_android_LCBInvocationHandler_d
-        extern void co_yield_to_android_and_call(co_yield_callback_t callback, void *context);
+        extern void co_yield_to_engine_and_call(co_yield_callback_t callback, void *context);
         remote_call_t t_context = {p_handler, p_method_name, p_args};
-        co_yield_to_android_and_call(remote_call_func, &t_context);
```
Isn't there a problem here with local JNI refs? The easiest way to fix this for now is to take GlobalRefs before doing the yield, and then delete them after the yield.

livecodeali replied:
Ah yes, I see what you mean

runrevmark commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -554,8 +554,11 @@ non-Windows platform then it is taken to be `default`.
 
-    "java:[className>][functionType.]function[!calling]"
-    
+    "java[thread]:[className>][functionType.]function[!calling]"
```
Is it worth considering putting the thread affinity somewhere else? Perhaps a final ?<thread-affinity> clause at the end.

livecodeali replied:
Yeah, I guess it probably is

livecodeali committed:
[[ ListenerCallbacks ]] Global-ref jobjs before jumping threads in native callback
The incoming jobjects will only be valid local refs for the
incoming JNIEnv, so ensure that the refs persist across the thread
jump by creating and deleting global refs around the call

livecodeali commented:
Depends on https://github.com/livecode/livecode/pull/5660 - will need rebasing once that is merged.

runrevmark commented:
@livecode-vulcan review ok 73fb7f1

livecode-vulcan commented:
💙 review by @runrevmark ok 73fb7f1dacd37b1a5eb736271f6e73cdf90232e1

livecode-vulcan commented:
😞 test failure 73fb7f1dacd37b1a5eb736271f6e73cdf90232e1

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/446)
* try-community-armv6-android-api9: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/454)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/433)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/376)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/376)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/441)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/446)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/444)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/454)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/444)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/453)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/471)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/440)

livecodeali committed:
[[ Bug 19911 ]] Fix android build error

livecodeali commented:
@runrevmark Fixed that build error.

runrevmark commented:
@livecode-vulcan review ok e4fc525

livecode-vulcan commented:
💙 review by @runrevmark ok e4fc5256f31d86b8b88fc6961b3ae59ad29c8fde

livecode-vulcan commented:
😎 test success e4fc5256f31d86b8b88fc6961b3ae59ad29c8fde

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/450)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/458)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/437)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/380)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/380)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/445)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/450)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/448)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/461)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/448)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/457)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/475)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/444)
